### PR TITLE
Change all messages to pass back their checksum. Switch to callback style

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Extracted from https://github.com/dotnet/roslyn/pull/73013 to help keep that PR simpler.

### CODE REVIEW

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 1]
```diff
@@ -39,33 +39,25 @@ internal static class RemoteHostAssetSerialization
...
+                var kind = asset.GetWellKnownSynchronizationKind();
+                checksum.WriteTo(writer);
+                writer.WriteInt32((int)kind);
+                serializer.Serialize(asset, writer, context, cancellationToken);
 
                 // We flush after each item as that forms a reasonably sized chunk of data to want to then send over the
                 // pipe for the reader on the other side to read.  This allows the item-writing to remain entirely
                 // synchronous without any blocking on async flushing, while also ensuring that we're not buffering the
                 // entire stream of data into the pipe before it gets sent to the other side.
-                WriteAsset(writer, serializer, context, asset, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:51]: inliend this method.

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 2]
```diff
@@ -307,22 +315,5 @@ Task SynchronizeProjectAssetCollectionAsync<TAsset>(AssetPath assetPath, Func<Pr
                     s_checksumPool.Free(missingChecksums);
             }
         }
-
-        return;
-    }
-
-    private async ValueTask RequestAssetsAsync<T, TArg>(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:314]: inlined this method.

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 3]
```diff
@@ -284,18 +284,26 @@ Task SynchronizeProjectAssetCollectionAsync<TAsset>(AssetPath assetPath, Func<Pr
                 if (missingChecksumsCount > 0)
                 {
                     var missingChecksumsMemory = new ReadOnlyMemory<Checksum>(missingChecksums, 0, missingChecksumsCount);
+                    Contract.ThrowIfTrue(missingChecksumsMemory.Length == 0);
 
-                    await RequestAssetsAsync(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:288]: inlined this method call.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**